feat: add scripting v0.8 blog post proposal (#290)
arcus-azure/arcus: community/blog/2023-01 Role-based authentication in Azure API Management made easy in Arcus Scripting v0.8/README.md

# Role-based authentication in Azure API Management made easy in Arcus Scripting v0.8
Managing roles during JWT validation was tedious and repetitive work. Luckily, the newest Arcus Scripting release changed that.

## Validating JSON web tokens in Azure API Management

> The `validate-jwt` policy in Azure API Management provides the capability to enforce a valid JWT in an incoming HTTP request. One of these validation rules is whether certain roles have been assigned to a service principal. This authorization functionality is very useful to allow/deny certain functionality to users of the application.

Reviewer: I would change the 'certain roles have been assigned to a service principal part' to something like this:

```suggestion
The `validate-jwt` policy in Azure API Management provides the capability to enforce a valid JWT in an incoming HTTP request. One of these validation rules is whether the received token contains certain role claims. This authorization functionality is very useful to allow/deny certain functionality to users of the application based on their role membership.
```

> The `validate-jwt` policy in Azure API Management provides the capability to consumers to enforce a valid JWT in an incoming HTTP request. One of these validation rules is whether certain roles have been assigned to a service principal. This authorization functionality is very useful to allow/deny certain functionality to users of the application.

Reviewer 2: ```suggestion
The `validate-jwt` policy in Azure API Management provides the capability to enforce a valid JWT in an incoming HTTP request. One of these validation rules is whether certain roles have been assigned to a service principal. This authorization functionality is very useful to allow/deny certain functionality to users of the application.
```

Management of service principal roles to certain Azure Directory Applications is rather tedious. One has to look up the role assignments of an Azure Active Directory application to find out if the service principal has the correct access. Moreover, in certain scenarios, one has to wait a couple of seconds before a role assignment is available for use. All this adds to the problem of managing a service principal for validating JWTs in Azure API Management.

An example of such policy is shown here:

> ```xml

Reviewer 2: I would move this part below `The validate-jwt policy in Azure API Management provides the capability to consumers to enforce a valid JWT in an incoming HTTP request. One of these validation rules is whether certain roles have been assigned to a service principal. This authorization functionality is very useful to allow/deny certain functionality to users of the application.` part

Author: Moved `validate-jwt` example below 1st paragraph in https://github.com/arcus-azure/arcus/pull/290/commits/36768f2ba60ffd9670f8d6e3bc89247c3a7aa6e7

    <issuers>
        <issuer>{{issuer}}</issuer>
    </issuers>
    <required-claims>
        <claim name="roles" match="any">
            <value>DummyRole</value>
        </claim>
    </required-claims>
</validate-jwt>
```

> Management of service principal roles to certain Azure Directory Applications is rather tedious. One has to look up the role assignments of an Azure Active Directory application to find out if the service principal has the correct access. Moreover, in certain scenarios, one has to wait a couple of seconds before a role assignment is available for use. All this adds to the problem of managing a service principal for validating JWTs in Azure API Management.

Reviewer: Here we're talking about a certain problem, but what is the solution ?

Author: Our PowerShell module 😄

Reviewer: Ok, but I meant that this is not obvious from the text :)

Author: But, we have **In a single Arcus Scripting release, we have fully fixed the problem of role assignments for a service principal**, right? In the next paragraph?

Reviewer: Maybe we should consider moving that one sentence from that paragraph to this paragraph ?

Author: Done in https://github.com/arcus-azure/arcus/pull/290/commits/e770ba5cdbda61bc7ab54ddb305ab9e99ad3729b

## List, add, and remove role assignments for a service principal

> In a single Arcus Scripting release, we have fully fixed the problem of role assignments for a service principal. In a new `Arcus.Scripting.ActiveDirectory` PowerShell module, we have created three functions that let you list, add, and remove role assignments to let a service principal access an Azure Active Directory.

Reviewer: ```suggestion
In a single Arcus Scripting release, we have fully fixed the problem of role assignments for a service principal. In a new `Arcus.Scripting.ActiveDirectory` PowerShell module, we have created three functions that let you list, add, and remove role assignments to a service principal in Azure Active Directory.
```

#Found role 'FirstRole' on Active Directory Application 'main-application'
#Role 'FirstRole' is assigned to the Active Directory Application 'client-application-one' with id '6ea09bbd-c21c-460c-b58a-f4a720f51826'
#Found role 'SecondRole' on Active Directory Application 'main-application'
#Role 'SecondRole' is assigned to the Active Directory Application 'client-application-one' with id '6ea09bbd-c21c-460c-b58a-f4a720f51826'

PS> Add-AzADAppRoleAssignment `
  -ClientId "b885c208-6067-44bd-aba9-4010c62b7d85" `
  -Role "DummyRole" `
  -AssignRoleToClientId "6ea09bbd-c21c-460c-b58a-f4a720f51826"
#Active Directory Application 'main-application' does not contain the role 'DummyRole', adding the role
#Added Role 'DummyRole' to Active Directory Application 'main-application'
#Role Assignment for the role 'DummyRole' added to the Active Directory Application 'client-application-one'

PS> Remove-AzADAppRoleAssignment `
  -ClientId "b885c208-6067-44bd-aba9-4010c62b7d85" `
  -Role "DummyRole" `
  -RemoveRoleFromClientId "6ea09bbd-c21c-460c-b58a-f4a720f51826" `
#Role assignment for 'DummyRole' has been removed from Active Directory Application 'client-application-one'

> ```

Reviewer: Playing the devil's advocate here, but why would I use these Arcus scripts instead of the native Azure CLI commands:
https://learn.microsoft.com/en-us/cli/azure/role/assignment?view=azure-cli-latest#az-role-assignment-create
 ?

Author: @pim-simons will be better suited to answer this, but as I understand it, it is about how our scripts are more fail-safe as they they into account existing roles and has a 'return until it's available' feature so that after the script has run, you know that the role assignment is ready to use.

Reviewer 2: Correct, and it combines several tasks into one script (creating the role and the role assignment).
